pkg/gadget-service/service.go

```diff
@@ -0,0 +1,291 @@
+		Catalog: catalogJSON,
+	}, nil
+}
+
+func (s *Service) RunGadget(runGadget pb.GadgetManager_RunGadgetServer) error {
+	ctx, cancel := context.WithCancel(runGadget.Context())
+	defer cancel()
+
+	ctrl, err := runGadget.Recv()
+	if err != nil {
+		return err
+	}
+
+	request := ctrl.GetRunRequest()
+	if request == nil {
+		return fmt.Errorf("expected first control message to be gadget request")
+	}
+
+	// Create a new logger that logs to gRPC and falls back to the standard logger when it failed to send the message
+	logger := logger.NewFromGenericLogger(&Logger{
+		send:           runGadget.Send,
+		level:          logger.Level(request.LogLevel),
+		fallbackLogger: s.logger,
+	})
+
+	runtime := s.runtime
+	defer func() {
+		// Try to send a done message
+		runGadget.Send(&pb.GadgetEvent{Type: pb.EventTypeGadgetDone})
+	}()
+
+	gadgetDesc := gadgetregistry.Get(request.GadgetCategory, request.GadgetName)
+	if gadgetDesc == nil {
+		logger.Errorf("gadget not found: %s/%s", request.GadgetCategory, request.GadgetName)
+		return nil
+	}
+
+	// TODO: Remove
+	logger.Debugf("Params")
+	for k, v := range request.Params {
+		logger.Debugf("- %s: %q", k, v)
+	}
+
+	// Initialize Operators
+	err = operators.GetAll().Init(operators.GlobalParamsCollection())
+	if err != nil {
+		logger.Errorf("initialize operators: %v", err)
+		return err
+	}
+
+	ops := operators.GetOperatorsForGadget(gadgetDesc)
+
+	operatorParams := ops.ParamCollection()
+	err = operatorParams.CopyFromMap(request.Params, "operator.")
+	if err != nil {
+		logger.Errorf("setting operator parameters: %v", err)
+		return nil
+	}
+
+	parser := gadgetDesc.Parser()
+
+	runtimeParams := runtime.ParamDescs().ToParams()
+	err = runtimeParams.CopyFromMap(request.Params, "runtime.")
+	if err != nil {
+		logger.Errorf("setting runtime parameters: %v", err)
+		return nil
+	}
+
+	gadgetParamDescs := gadgetDesc.ParamDescs()
+	gadgetParamDescs.Add(gadgets.GadgetParams(gadgetDesc, parser)...)
+	gadgetParams := gadgetParamDescs.ToParams()
+	err = gadgetParams.CopyFromMap(request.Params, "")
+	if err != nil {
+		logger.Errorf("setting gadget parameters: %v", err)
+		return nil
+	}
+
+	// Create payload buffer
+	outputBuffer := make(chan *pb.GadgetEvent, 1024) // TODO: Discuss 1024
+	outputDone := make(chan bool)
+	defer func() {
+		outputDone <- true
+	}()
+	seq := uint32(0)
+	var seqLock sync.Mutex
+
+	if parser != nil {
+		parser.SetLogCallback(logger.Logf)
+		parser.SetEventCallback(func(ev any) {
+			// Marshal messages to JSON
+			// Normally, it would be better to have this in the pump below rather than marshaling events that
+			// would be dropped anyway. However, we're optimistic that this occurs rarely and instead prevent using
+			// ev in another thread.
+			data, _ := json.Marshal(ev)
+			lev := &pb.GadgetEvent{
+				Type:    pb.EventTypeGadgetPayload,
+				Payload: data,
+			}
+
+			seqLock.Lock()
+			seq++
+			lev.Seq = seq
+			select {
+			case outputBuffer <- lev:
+			default:
+			}
```

Comment:
Sure - I'll add a couple of comments as well:

The real output to the gRPC connection happens in the go routine below, but the `runGadget.Send(ev)` is blocking. So if your client is slow / has a network drop, Send() will fill up the connection buffers (there's one from the OS for tcp connections and another one provided by the gRPC implementation) and then stall until space becomes available. If we were to just forward the events directly, the whole gadget could stall in such a case. That's why we use the channel (`outputBuffer`) as an additional buffer - using `select` with one option being to send something through a channel, and the other being `default`, makes it so that if `outputBuffer` is full (and therefore would block), it would instead select the `default` case. In this example that leads to just dropping the event.